Ignore deleting vmimage in storage class validator

brandboat opened the pull request:
#### Problem:

StorageClass validator blocks the storage class deletion even if its related vm image is marked as deleting.

This is a problem when we perform the following operation sequentially:
1. delete vmimage
2. delete the storage class that relates to the vmimage

In step 1 k8s marks vmimage as deleting, but the vm image has not been fully removed yet, then step 2 fails because the storage class validate treat it as still in-use.

#### Solution:

ignore deleting vm images in storage class validator

#### Related Issue(s):
https://github.com/harvester/harvester/issues/9720

#### Test plan:
well-described in issue description, see https://github.com/harvester/harvester/issues/9720

https://github.com/user-attachments/assets/26833d86-7d30-48de-aa70-339edc0c9186

#### Additional documentation or context

brandboat committed:
Ignore deleting vmimage in storage class validator

This allows us to delete a storage class even when a VM image is marked
as deleting but is still associated with the target storage class.

Signed-off-by: Cooper Tseng <cooper.tseng@suse.com>

Yu-Jack reviewed (approved):
Thanks!

WebberHuang1118 commented:
Hi @brandboat,

In my opinion, we should not change the fundamental behavior of the StorageClass validator. Since there is still a VMI image referencing the target StorageClass, blocking the deletion is reasonable.

If I understand correctly, in this case the StorageClass deletion is triggered by the VMI controller. We could handle this webhook error as a special case and add a retry mechanism.

Thanks.

brandboat commented on pkg/webhook/resources/storageclass/validator.go:
```diff
@@ -322,7 +322,9 @@ func (v *storageClassValidator) validateVMImageUsage(sc *storagev1.StorageClass)
-		usedVMImages = append(usedVMImages, vmimage.Name)
+		if vmimage.DeletionTimestamp == nil {
+			usedVMImages = append(usedVMImages, vmimage.Name)
+		}
```
Actually the storage class deletion is triggered by UI, see https://github.com/harvester/harvester-ui-extension/blob/main/pkg/harvester/pages/c/_cluster/airgapupgrade/index.vue#L364-L380

brandboat commented:
> Why only the upgrade image has such situation?

This is closely related to deprecating the upgrade-repo VM (see https://github.com/harvester/harvester/issues/7101
). To make the upgrade repo more robust, we switched from a VM-based approach to a deployment-based one. As part of this change, the underlying PVC of the VM image must use filesystem mode with RWX access, that's why we have this special handling for upgrade images.

> The sc is deleted by front-end for the upgrade case?

There are two different upgrade flows in Harvester:

1. Standard upgrade flow (via the Version CR and Upgrade CR):
In this case, the temporary storage class is created by the backend and is eventually cleaned up by the backend as well.
2. Air-gap upgrade flow (the scenario this PR aims to fix):
This flow is more complex. The VM image is uploaded directly by the user rather than being created by the backend. As a result, the temporary storage class is created by the UI and is also deleted by the UI. This is because the VMImage’s `.spec.targetStorageClassName` must be populated upfront, which means the storage class has to be created before the VM image is created.

WebberHuang1118 commented:
Thanks for the explanation, @brandboat.

> Regarding https://github.com/harvester/harvester/pull/9785#discussion_r2650145902 One potential risk of this PR is that if deleting VMI images actually not removed, e.g. UI trigger VMI deletion & SC deletion, while VMI deletion operation stuck, user accidentally remove deletion timestamp, then VMI won't be deleted eventually.

Based on the discussions so far, I think the current StorageClass validator design is reasonable. The StorageClass should not be deleted until the related VM is fully removed, so we may not need to change this behavior in this PR.

> Back to the alternative solution discussion, there are 2 alternative ways to do this
> 1. proposed in comment https://github.com/harvester/harvester/pull/9785#issuecomment-3698163637
> 2. wait until VMI deletion completes then do SC deletion in the UI side.
> 

I prefer option 2, as it’s more rational from an overall design perspective. If this isn’t feasible on the front end, option 1 would also be acceptable to me.

brandboat commented:
> (3) The backend change, per your PR, are there any labels/annotations to identify this air-gapped upgrade related vmimage & SC ? if we can identify it, then relax the SC webhook check to ignore the vmimage with deletiontimestap might be a good balance.

@w13915984028, yes, the annotation is here:`harvesterhci.io/os-upgrade-image`, I'll go this way - relax the sc deletion  webhook

w13915984028 commented:
@brandboat It looks the `longhorn-static` SC is a bit different with the default SC `harvester-longhorn`, both the labels/annotations and those normal parameters, are we sure it can be used for this case?

UI upgrade seems to have no option to set param like replica-count
<img width="3212" height="1140" alt="image" src="https://github.com/user-attachments/assets/317e2207-1691-488f-9e5c-6beafd0e302c" />

but a normal image has such param:
<img width="2750" height="722" alt="image" src="https://github.com/user-attachments/assets/f8dc132a-27e3-4e9f-bca9-affd5baf8317" />



chart doc also says:

>deploy/charts/harvester/charts/longhorn/questions.yaml:          Default name of Longhorn Static StorageClass. The "storageClassName" is assigned to PVs and PVCs that are created for an existing Longhorn volume. The "storageClassName" can also be used as a label, so it is possible to use a Longhorn StorageClass to bind a workload to an existing PV without creating a Kubernetes StorageClass object. The "storageClassName" needs to be an existing StorageClass. The default value is "longhorn-static".



```
$kubectl get storageclass longhorn-static -oyaml
allowVolumeExpansion: true
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  creationTimestamp: "2025-11-12T09:20:29Z"
  name: longhorn-static
  resourceVersion: "5647"
  uid: caa3e394-1afc-47de-95aa-61341168eb81
parameters:
  staleReplicaTimeout: "30"
provisioner: driver.longhorn.io
reclaimPolicy: Delete
volumeBindingMode: Immediate


$kubectl get storageclass harvester-longhorn -oyaml
allowVolumeExpansion: true
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  annotations:
    harvesterhci.io/is-reserved-storageclass: "true"
    meta.helm.sh/release-name: harvester
    meta.helm.sh/release-namespace: harvester-system
    objectset.rio.cattle.io/id: default-mcc-harvester-cattle-fleet-local-system
    storageclass.kubernetes.io/is-default-class: "true"
  creationTimestamp: "2025-11-12T09:19:54Z"
  labels:
    app.kubernetes.io/managed-by: Helm
    objectset.rio.cattle.io/hash: e852fa897f5eae59a44b4bfe186aad80b10b94b3
  name: harvester-longhorn
  resourceVersion: "3743"
  uid: f92f9629-700e-4a36-88f4-bd3aadea833e
parameters:
  baseImage: ""
  fromBackup: ""
  migratable: "true"
  numberOfReplicas: "3"
  staleReplicaTimeout: "30"
provisioner: driver.longhorn.io
reclaimPolicy: Delete
volumeBindingMode: Immediate
```

w13915984028 commented:
@brandboat Another case arise:

The `longhorn-static` SC is an OOB SC from the view of Harvester (not installed & managed via helm chart, no helm labels/annotations), on UI, it has a hint that it is internal and can't be deleted.

<img width="1604" height="706" alt="image" src="https://github.com/user-attachments/assets/26d53979-50f4-40e6-b4fd-29632ddad951" />


But, if I delete `longhorn-static` SC from `kubectl`, it just runs smoothly. Webhook does not deny it, as `longhorn-static` does not have the `harvesterhci.io/is-reserved-storageclass: "true"` annotation like `harvester-longhorn`.

https://github.com/harvester/harvester/blob/dedd291efd8c659a9f05e13b62b615477e708993/pkg/webhook/resources/storageclass/validator.go#L335-L357

```
harv41:/home/rancher # kubectl get storageclass -A
NAME                           PROVISIONER          RECLAIMPOLICY   VOLUMEBINDINGMODE   ALLOWVOLUMEEXPANSION   AGE
harvester-longhorn (default)   driver.longhorn.io   Delete          Immediate           true                   62d
longhorn                       driver.longhorn.io   Delete          Immediate           true                   62d
longhorn-image-666qx           driver.longhorn.io   Delete          Immediate           true                   57d
longhorn-image-f2vpr           driver.longhorn.io   Delete          Immediate           true                   56d
longhorn-image-fzk7j           driver.longhorn.io   Delete          Immediate           true                   61d
longhorn-image-p8vl4           driver.longhorn.io   Delete          Immediate           true                   56d
longhorn-static                driver.longhorn.io   Delete          Immediate           true                   62d
vmstate-persistence            driver.longhorn.io   Delete          Immediate           true                   62d
harv41:/home/rancher # kubectl delete storageclass longhorn-static
storageclass.storage.k8s.io "longhorn-static" deleted
harv41:/home/rancher # kubectl get storageclass -A
NAME                           PROVISIONER          RECLAIMPOLICY   VOLUMEBINDINGMODE   ALLOWVOLUMEEXPANSION   AGE
harvester-longhorn (default)   driver.longhorn.io   Delete          Immediate           true                   62d
longhorn                       driver.longhorn.io   Delete          Immediate           true                   62d
longhorn-image-666qx           driver.longhorn.io   Delete          Immediate           true                   57d
longhorn-image-f2vpr           driver.longhorn.io   Delete          Immediate           true                   56d
longhorn-image-fzk7j           driver.longhorn.io   Delete          Immediate           true                   61d
longhorn-image-p8vl4           driver.longhorn.io   Delete          Immediate           true                   56d
vmstate-persistence            driver.longhorn.io   Delete          Immediate           true                   62d
harv41:/home/rancher # 
```

By nature, we can assume that some customers remove it. If UI tries to refer it, another issue could happen.

Another point, to decouple with `longhorn`, Harvester adds `harvester-longhorn` as the default SC from installation;  from source code, we do not allow to refer `longhorn-static` but has no protection upon it.  Even if we want to refer kind of pre-existing SC, the `harvester-longhorn` looks better than `longhorn-static`.


Again, what's the consideration for UI to implement a new OS-image rather than reuse the original image creation process? It seems to import some normal/corner error cases.

<img width="2364" height="972" alt="image" src="https://github.com/user-attachments/assets/ebf87d3a-2876-4b7c-97e9-359e461b58b5" />

w13915984028 commented:
Per above code, I set this scenario:

1. create new SC longhorn-static1
2. edit longhorn setting default-longhorn-static-storage-class  to above sc
3. delete `longhorn-static`
4. it won't be recreated

I guess the `migratable: "true"` can even be set to `longhorn-static1`

using `longhorn-static` from current implementation, have a few flaws

the uncontrolled SC, can be none-existing, or with unwanted params

```
cat > lh1.yaml << 'EOF'
allowVolumeExpansion: true
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: longhorn-static1
parameters:
  staleReplicaTimeout: "30"
provisioner: driver.longhorn.io
reclaimPolicy: Delete
volumeBindingMode: Immediate
EOF


kubectl get storageclass longhorn-static1  -oyaml
allowVolumeExpansion: true
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  creationTimestamp: "2026-01-14T09:42:39Z"
  name: longhorn-static1
  resourceVersion: "4002766"
  uid: f8056255-a086-4e48-8127-3dfd63aebbe3
parameters:
  staleReplicaTimeout: "30"
provisioner: driver.longhorn.io
reclaimPolicy: Delete
volumeBindingMode: Immediate

kubectl get settings.longhorn -n longhorn-system default-longhorn-static-storage-class -oyaml
apiVersion: longhorn.io/v1beta2
kind: Setting
metadata:
  annotations:
    longhorn.io/configmap-resource-version: "3740"
  creationTimestamp: "2025-11-12T09:20:27Z"
  generation: 2
  name: default-longhorn-static-storage-class
  namespace: longhorn-system
  resourceVersion: "4002924"
  uid: 9a1ebac7-e767-46eb-b7b6-5d980db64983
status:
  applied: true
value: longhorn-static1

kubectl get storageclass -A
NAME                           PROVISIONER          RECLAIMPOLICY   VOLUMEBINDINGMODE   ALLOWVOLUMEEXPANSION   AGE
harvester-longhorn (default)   driver.longhorn.io   Delete          Immediate           true                   63d
longhorn                       driver.longhorn.io   Delete          Immediate           true                   63d
longhorn-image-666qx           driver.longhorn.io   Delete          Immediate           true                   57d
longhorn-image-f2vpr           driver.longhorn.io   Delete          Immediate           true                   56d
longhorn-image-fzk7j           driver.longhorn.io   Delete          Immediate           true                   61d
longhorn-image-p8vl4           driver.longhorn.io   Delete          Immediate           true                   56d
longhorn-static1               driver.longhorn.io   Delete          Immediate           true                   2m50s
vmstate-persistence            driver.longhorn.io   Delete          Immediate           true                   63d

```
